tests/test_data.py

```diff
@@ -1259,6 +1259,92 @@ def test_misc_data(self):
             assert(np.all(c.data[-1] == [4., 4., 6., 9., 8.]))
 
 
+class TestDataGather(object):
+
+    @pytest.mark.parallel(mode=4)
+    @pytest.mark.parametrize('rank', [0, 1, 2, 3])
+    def test_simple_gather(self, rank):
+        """ Test a simple gather on various ranks."""
+        grid = Grid(shape=(10, 10), extent=(9, 9))
+        f = Function(name='f', grid=grid, dtype=np.int32)
+        res = np.arange(100).reshape(grid.shape)
+        f.data[:] = res
+        myrank = grid._distributor.comm.Get_rank()
+        ans = f.data_gather(rank=rank)
+        if myrank == rank:
+            assert np.all(ans == res)
+        else:
+            assert ans.shape == (0, )*len(grid.shape)
+
+    @pytest.mark.parallel(mode=4)
+    @pytest.mark.parametrize('start, stop, step', [
+        (None, None, None),
+        (None, None, 2),
+        (None, None, -1),
+        (None, None, -2),
+        (1, 8, 3),
+        ((0, 4), None, (2, 1))])
+    def test_sliced_gather_2D(self, start, stop, step):
+        """ Test gather for various 2D slices."""
+        grid = Grid(shape=(10, 10), extent=(9, 9))
+        f = Function(name='f', grid=grid, dtype=np.int32)
+        dat = np.arange(100).reshape(grid.shape)
+
+        if isinstance(step, int) or step is None:
+            step = [step for _ in grid.shape]
+        if isinstance(start, int) or start is None:
+            start = [start for _ in grid.shape]
+        if isinstance(stop, int) or stop is None:
+            stop = [stop for _ in grid.shape]
+        idx = []
+        for i, j, k in zip(start, stop, step):
+            idx.append(slice(i, j, k))
+        idx = tuple(idx)
+
+        res = dat[idx]
+        f.data[:] = dat
+        myrank = grid._distributor.comm.Get_rank()
+        ans = f.data_gather(start=start, stop=stop, step=step)
+        if myrank == 0:
+            assert np.all(ans == res)
+        else:
+            assert ans.shape == (0, )*len(grid.shape)
+
+    @pytest.mark.parallel(mode=4)
+    @pytest.mark.parametrize('start, stop, step', [
+        (None, None, None),
+        (None, None, 2),
+        (None, None, -1),
+        (None, None, -2),
+        (1, 8, 3),
+        ((0, 4, 4), None, (2, 1, 1))])
+    def test_sliced_gather_3D(self, start, stop, step):
+        """ Test gather for various 2D slices."""
```

Comment:
3D 😬

Reply:
oh, bleh, thanks. I'll add this and the thing @FabioLuporini mentioned to #1457.